torchmultimodal/modules/layers/image_embedding.py

```diff
@@ -0,0 +1,163 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+import warnings
+from typing import Optional
+
+import torch
+from torch import nn, Tensor
+
+
+# Based on timm implementation, which can be found here:
+# https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py
+class PatchEmbeddings(nn.Module):
+    """
+    Image to Patch Embedding.
+    """
+
+    def __init__(
+        self,
+        image_size: int = 224,
+        patch_size: int = 16,
+        num_channels: int = 3,
+        embed_dim: int = 768,
+    ) -> None:
+        super().__init__()
+        image_size = (image_size, image_size)
+        patch_size = (patch_size, patch_size)
+        num_patches = (image_size[1] // patch_size[1]) * (
+            image_size[0] // patch_size[0]
+        )
+        self.image_size = image_size
+        self.patch_size = patch_size
+        self.num_patches = num_patches
+
+        self.projection = nn.Conv2d(
+            num_channels, embed_dim, kernel_size=self.patch_size, stride=self.patch_size
+        )
+
+    def forward(
+        self, pixel_values: Tensor, interpolate_pos_encoding: bool = False
+    ) -> Tensor:
+        _, _, height, width = pixel_values.shape
+        if not interpolate_pos_encoding:
+            if height != self.image_size[0] or width != self.image_size[1]:
+                raise ValueError(
+                    f"Input image size ({height}*{width}) doesn't match model ({self.image_size[0]}*{self.image_size[1]})."
+                )
+        x = self.projection(pixel_values).flatten(2).transpose(1, 2)
+        return x
+
+
+class ImageEmbeddings(nn.Module):
+    """
+    Construct the CLS token, position and patch embeddings.
+    """
+
+    def __init__(
+        self,
+        image_size: int = 224,
+        patch_size: int = 16,
+        num_channels: int = 3,
+        hidden_size: int = 768,
+        hidden_dropout_prob: float = 0.0,
+        use_image_masking: bool = True,
+    ) -> None:
+        super().__init__()
+
+        self.cls_token = nn.Parameter(torch.zeros(1, 1, hidden_size))
+        self.patch_embeddings = PatchEmbeddings(
+            image_size=image_size,
+            patch_size=patch_size,
+            num_channels=num_channels,
+            embed_dim=hidden_size,
+        )
+        num_patches = self.patch_embeddings.num_patches
+        self.position_embeddings = nn.Parameter(
+            torch.zeros(1, num_patches + 1, hidden_size)
+        )
+        self.dropout = nn.Dropout(hidden_dropout_prob)
+
+        if use_image_masking:
+            self.mask_token = nn.Parameter(torch.zeros(1, 1, hidden_size))
+        else:
+            self.mask_token = None
+
+    def interpolate_pos_encoding(
```

Comment:
What's the use case for this? I see it's added as an option for FLAVA but don't see it ever being used. cc @ankitade